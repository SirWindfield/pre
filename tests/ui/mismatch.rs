#![feature(proc_macro_hygiene)]
#![feature(stmt_expr_attributes)]

use pre::{assert_precondition, pre};

#[pre(condition(valid_ptr(ptr)), condition("must point to `42`"))]
unsafe fn read_twice<T: Copy>(ptr: *const T) -> (T, T) {
    (std::ptr::read(ptr), std::ptr::read(ptr))
}

fn main() {
    let ptr: *const i32 = &42;

    let (_, _) = unsafe {
        #[assert_precondition(holds(valid_ptr(ptr), reason = "`ptr` comes from a reference"))]
        read_twice(ptr)
    };
}
