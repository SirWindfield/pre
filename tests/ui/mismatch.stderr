error[E0308]: mismatched types
  --> $DIR/mismatch.rs:15:9
   |
15 |         #[assert_precondition(holds(valid_ptr(ptr), reason = "`ptr` comes from a reference"))]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected a tuple with 2 elements, found one with 1 element
   |
   = note: expected struct `std::marker::PhantomData<(pre::const_generics_types::ValidPtrConditionHolds<"ptr">, pre::const_generics_types::CustomConditionHolds<"must point to `42`">)>`
              found struct `std::marker::PhantomData<(pre::const_generics_types::ValidPtrConditionHolds<"ptr">,)>`
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
